### PR TITLE
Estimate gas price

### PR DESCRIPTION
**Pull request by zachzundel**
So far, I have the gas price being queried from the API if the default is being used. I still need to fix the tests and maybe add a `--gasPrice` flag to the CLI. This is for issue #56 .

### CODE REVIEW

**Comment by zachzundel** [comment]
@spalladino Could you take a look at this and let me know if it seems reasonable so far?

**Comment by spalladino** [comment]
@zachzundel looks good so far! A few comments:

- Try to cache the gas price obtained from the API in the `state` object, to avoid having to wait one query per each tx.
- Double-check the units in which the API returns the gas price, you may need to add a few zeroes to the value you obtain.
- Make sure to add a few tests later down the road (I know this is still WIP).
- And a minor comment: these lines are repeated several times, see if it's possible to refactor them:
```js
   let fixedGasPrice = checkGasPrice(txParams)
   if (fixedGasPrice) {
      txParams.gasPrice = fixedGasPrice
  }
```

**Comment by spalladino** [comment]
@zachzundel looks good so far! A few comments:

- Try to cache the gas price obtained from the API in the `state` object, to avoid having to wait one query per each tx.
- Double-check the units in which the API returns the gas price, you may need to add a few zeroes to the value you obtain.
- Make sure to add a few tests later down the road (I know this is still WIP).
- And a minor comment: these lines are repeated several times, see if it's possible to refactor them:
```js
   let fixedGasPrice = checkGasPrice(txParams)
   if (fixedGasPrice) {
      txParams.gasPrice = fixedGasPrice
  }
```

**Comment by spalladino** [comment]
@zachzundel looks good so far! A few comments:

- Try to cache the gas price obtained from the API in the `state` object, to avoid having to wait one query per each tx.
- Double-check the units in which the API returns the gas price, you may need to add a few zeroes to the value you obtain.
- Make sure to add a few tests later down the road (I know this is still WIP).
- And a minor comment: these lines are repeated several times, see if it's possible to refactor them:
```js
   let fixedGasPrice = checkGasPrice(txParams)
   if (fixedGasPrice) {
      txParams.gasPrice = fixedGasPrice
  }
```

**Comment by spalladino** [comment]
@zachzundel looks good so far! A few comments:

- Try to cache the gas price obtained from the API in the `state` object,
to avoid having to wait one query per each tx.
- Double-check the units in which the API returns the gas price, you may
need to add a few zeroes to the value you obtain.
- Make sure to add a few tests later down the road (I know this is still
WIP).
- And a minor comment: these lines are repeated several times, see if it's
possible to refactor them:
```js
   let fixedGasPrice = checkGasPrice(txParams)
   if (fixedGasPrice) {
      txParams.gasPrice = fixedGasPrice
  }
```

(sorry if you see this comment several times, I'm getting an error when
posting it)

**Comment by zachzundel** [comment]
Hi @spalladino -- I've updated my PR and I think I've fixed everything you pointed out. There was already a test for estimateGas that seems to cover my use case, please let me know if that's not correct and I will write another one!

File: packages/lib/src/utils/Transactions.js [hunk 1]
```diff
@@ -152,11 +156,28 @@ async function getNodeVersion () {
...
-    throw new Error(`Cowardly refusing to execute transaction with excessively high default gas price of 100 gwei. Consider explicitly setting a different gasPrice value in your truffle.js config file. You can check reasonable values for gas price in https://ethgasstation.info/.`);
+    if (state.gasPrice) {
+        txParams.gasPrice = state.gasPrice;
+        return txParams;
+    }
+
+    try {
+      let apiResponse = await axios.get(GAS_API_URL);
+      let gasPriceGwei = apiResponse.safeLow / 10;
+      let gasPrice = BN(gasPriceGwei * 1000000000)
```
[the remark this answers is not in the document]
Comment (zachzundel) [reply packages/lib/src/utils/Transactions.js]: Can do! Out of curiousity, can you point me to a writeup on why one should prefer const over let?

**Comment by zachzundel** [comment]
How can we tell testnet vs. mainnet here? Is it something in `txParams`? Do I need to use Truffle?

File: packages/lib/src/utils/Transactions.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (spalladino) [reply packages/lib/src/utils/Transactions.js]: Sure! From [ESLint](https://eslint.org/docs/rules/prefer-const):

> const declaration tells readers, “this variable is never reassigned,” reducing cognitive load and improving maintainability.

**Comment by zachzundel** [comment]
@spalladino I've added in tests, this is ready for review!

File: package.json [hunk 2]
```diff
@@ -3,5 +3,8 @@
   "private": true,
   "devDependencies": {
     "lerna": "~3.3.2"
+  },
+  "dependencies": {
+    "8": "0.0.1"
```
Comment (spalladino) [inline package.json]: What is this dependency for? Please remove if unneeded.

File: packages/lib/src/utils/Transactions.js [hunk 3]
```diff
@@ -152,12 +156,48 @@ async function getNodeVersion () {
...
+}
+
+async function fixGasPrice(txParams) {
+  const network = await getNetwork();
+
   const gasPrice = txParams.gasPrice || Contracts.artifactsDefaults().gasPrice;
-  if (TRUFFLE_DEFAULT_GAS_PRICE.eq(gasPrice) || !gasPrice) {
-    throw new Error(`Cowardly refusing to execute transaction with excessively high default gas price of 100 gwei. Consider explicitly setting a different gasPrice value in your truffle.js config file. You can check reasonable values for gas price in https://ethgasstation.info/.`);
+
+  if (TRUFFLE_DEFAULT_GAS_PRICE == gasPrice || !gasPrice) {
```
Comment (spalladino) [inline packages/lib/src/utils/Transactions.js]: Let's use bignumber equality functions for this, ie `TRUFFLE_DEFAULT_GAS_PRICE.eq(gasPrice)`

File: packages/lib/src/utils/Transactions.js [hunk 4]
```diff
@@ -152,12 +156,48 @@ async function getNodeVersion () {
...
+    if (network != '1') {
```
Comment (spalladino) [inline packages/lib/src/utils/Transactions.js]: We shouldn't need to do anything if network id is not mainnet, I think we can just leave gasPrice empty and let truffle pick the default. We can just `return` if `network != 1`.

File: packages/lib/src/utils/Transactions.js [hunk 5]
```diff
@@ -152,12 +156,48 @@ async function getNodeVersion () {
...
+      txParams.gasPrice = TRUFFLE_DEFAULT_GAS_PRICE.toNumber();
+      return txParams;
+    }
+
+    txParams.gasPrice = await getETHGasStationPrice()
+
+    if (txParams.gasPrice >= TRUFFLE_DEFAULT_GAS_PRICE) {
```
Comment (spalladino) [inline packages/lib/src/utils/Transactions.js]: Same here, let's use `TRUFFLE_DEFAULT_GAS_PRICE.lte(txParams.gasPrice)`

File: packages/lib/src/utils/Transactions.js [hunk 6]
```diff
@@ -32,7 +36,7 @@ const TRUFFLE_DEFAULT_GAS_PRICE = BN(100000000000);
  * @param retries number of transaction retries
  */
 export async function sendTransaction(contractFn, args = [], txParams = {}, retries = RETRY_COUNT) {
-  await checkGasPrice(txParams);
+  txParams = await fixGasPrice(txParams)
```
Comment (spalladino) [inline packages/lib/src/utils/Transactions.js]: `fixGasPrice` is both modifying `txParams` and returning it, which may be confusing to users of the function. Let's either: 1- remove the return value, and just call it as `await fixGasPrice(txParams)` (without the need for assigning to txParams)
2- or have fixGasPrice return a copy of txParams with gasPrice set (without modifying the parameter, so you'd need to make a copy of the txParams parameter among the first things you do in fixGasPrice)

File: packages/lib/src/utils/Transactions.js [hunk 7]
```diff
@@ -152,12 +156,48 @@ async function getNodeVersion () {
...
+  if (TRUFFLE_DEFAULT_GAS_PRICE == gasPrice || !gasPrice) {
+    if (network != '1') {
+      txParams.gasPrice = TRUFFLE_DEFAULT_GAS_PRICE.toNumber();
+      return txParams;
+    }
+
+    txParams.gasPrice = await getETHGasStationPrice()
+
+    if (txParams.gasPrice >= TRUFFLE_DEFAULT_GAS_PRICE) {
+        throw new Error(`Gas price API gave very high value (>100gwei), please manually provide a gas price.`)
```
Comment (spalladino) [inline packages/lib/src/utils/Transactions.js]: Let's rephrase to "The current gas price estimate from ethgasstation.info is over 100 gwei. If you do want to send a transaction with a gas price this high, please set it manually in your truffle.js configuration file."

File: packages/lib/src/utils/Transactions.js [hunk 8]
```diff
@@ -152,12 +156,48 @@ async function getNodeVersion () {
...
-  if (await isGanacheNode()) return;
+async function getNetwork() {
+  if (!state.network) {
+    state.network = await promisify(web3.version.getNetwork.bind(global.web3.version))();
+  }
+  return state.network;
+}
+
+async function getETHGasStationPrice() {
+    if (state.gasPrice) return state.gasPrice;
```
Comment (spalladino) [inline packages/lib/src/utils/Transactions.js]: Please use two spaces for indentation, for consistency

File: packages/lib/test/src/utils/Transactions.test.js [hunk 9]
```diff
@@ -67,6 +73,34 @@ contract('Transactions', function([_account1, account2]) {
       assertGasLt(tx, 1000000);
     });
 
+    it('uses gas price API when gas not specified', async function () {
+        const oldNetwork = state.network
```
Comment (spalladino) [inline packages/lib/test/src/utils/Transactions.test.js]: Please use two spaces for indentation, for consistency

File: packages/lib/test/src/utils/Transactions.test.js [hunk 10]
```diff
@@ -67,6 +73,34 @@ contract('Transactions', function([_account1, account2]) {
...
+
+        state.network = "1";
+        const stub = sinon.stub(axios, 'get').resolves({ average: 49 })
```
Comment (spalladino) [inline packages/lib/test/src/utils/Transactions.test.js]: Let's move these two lines to a beforeEach block. If needed, we can wrap these two tests in a "describe" block, so the beforeEach only affects them.

File: packages/lib/test/src/utils/Transactions.test.js [hunk 11]
```diff
@@ -67,6 +73,34 @@ contract('Transactions', function([_account1, account2]) {
...
+
+        const { tx } = await sendTransaction(this.instance.initialize, [42, 'foo', [1,2,3]]);
+      
+        assertGasPrice(tx, 49 * 1e8);
+
+        stub.restore()
+        state.network = oldNetwork
```
Comment (spalladino) [inline packages/lib/test/src/utils/Transactions.test.js]: Same here but with an afterEach

**Comment by zachzundel** [comment]
I've pushed commits to address all your changes!

**Pull request merged**